subsys/logging/log_backend_flash.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <logging/log_backend.h>
+#include "log_backend_std.h"
+#include <assert.h>
+#include <fs/fs.h>
+#include <fs/littlefs.h>
+
+#define MAX_PATH_LEN 255
+#define MAX_FLASH_WRITE_SIZE 256
+#define LOG_FILENAME_LEN (sizeof(CONFIG_LOG_FLASH_FILENAME) - 1)
+
+FS_LITTLEFS_DECLARE_DEFAULT_CONFIG(storage);
+static struct fs_mount_t lfs_storage_mnt = {
+	.type = FS_LITTLEFS,
+	.fs_data = &storage,
+	.storage_dev = (void *)FLASH_AREA_ID(storage),
+	.mnt_point = "/lfs",
+};
+
+static const char *log_filename = CONFIG_LOG_FLASH_FILENAME;
+static char fname[MAX_PATH_LEN];
+static struct fs_file_t file;
+static struct fs_statvfs stat;
+
+/* In case of no log file or current file is near overflow
+ * create new log file.
+ */
+static int allocate_new_file(struct fs_file_t *file);
+
+#if defined(CONFIG_OVERWRITE_OLD_LOG_FILES)
+static int delete_oldest_logs(void);
+#endif
+
+static int write_log_to_file(uint8_t *data, size_t length, void *ctx)
+{
+	int rc;
+	struct fs_file_t *f = (struct fs_file_t *)ctx;
+
+	assert(ctx);
+	int size = fs_tell(f);
+	/* Check if new data overwrites max file size.
+	 * If so, create new log file.
+	 */
+	if ((size + length) > (stat.f_frsize * CONFIG_LOG_FLASH_FILE_SIZE)) {
+		int ret = allocate_new_file(f);
+
+		if (ret) {
+			return length;
+		}
+	}
+	rc = fs_write(f, data, length);
+	if (rc == LFS_ERR_NOSPC) {
+#if defined(CONFIG_OVERWRITE_OLD_LOG_FILES)
```

Comment:
use `if (IS_ENABLED(CONFIG_OVERWRITE_OLD_LOG_FILES))` instead. This way ifdefed content is compiled. You can then remove ifdef from `delete_oldest_logs` declaration. Compiler will get rid of things which are determined to be unused during compilation.